From 7e8cfd13af5b12fa44c1020996be6439081ea4db Mon Sep 17 00:00:00 2001
From: Claudiu Manoil <claudiu.manoil@nxp.com>
Date: Tue, 14 Nov 2017 14:56:57 +0200
Subject: [PATCH 078/706] enetc: Fix TxBD flags handling for single BD frames

IE flag was unintentionally cleared for the case where the
first BD in a frame was also the last one (single BD frames).
Refactor 'flags' field from 16bit to 8bit to better reflect
the current ENETC BG.  'csum_err' was removed from the latest
ENETC BG versions, so drop it.
Minor refactoring to prepare addition of Tx offloads.

Signed-off-by: Claudiu Manoil <claudiu.manoil@nxp.com>
(cherry picked from commit 7e3e30b0280c3b6f496e0dd3beeeb40f94711abb)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c    |  8 +++++---
 drivers/net/ethernet/freescale/enetc/enetc_hw.h | 12 ++++++++----
 2 files changed, 13 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index 83b24437327a..8191ca8c0e8a 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -129,23 +129,25 @@ static void enetc_update_txbdr(struct enetc_bdr *tx_ring, int count, u16 len)
 {
 	struct enetc_tx_swbd *tx_swbd;
 	struct enetc_tx_bd *txbd;
+	u8 flags;
 	int i;
 
 	i = tx_ring->next_to_use;
 	txbd = ENETC_TXBD(*tx_ring, i);
 	tx_swbd = &tx_ring->tx_swbd[i];
 
+	flags = ENETC_TXBD_FLAGS_IE;
+
 	/* first BD needs frm_len set */
 	txbd->frm_len = cpu_to_le16(len);
-	txbd->flags = cpu_to_le16(ENETC_TXBD_FLAGS_IE);
 
 	while (count--) {
 		txbd->addr = cpu_to_le64(tx_swbd->dma);
 		txbd->buf_len = cpu_to_le16(tx_swbd->len);
 
 		/* last BD needs 'F' bit set */
-		if (!count)
-			txbd->flags = cpu_to_le16(ENETC_TXBD_FLAGS_F);
+		txbd->flags = count ? flags : flags | ENETC_TXBD_FLAGS_F;
+		flags = 0;
 
 		tx_swbd++;
 		txbd++;
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_hw.h b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
index 842e878a4e0b..265cafd2758d 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_hw.h
+++ b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
@@ -161,12 +161,16 @@ struct enetc_tx_bd {
 	__le64 addr;
 	__le16 buf_len;
 	__le16 frm_len;
-	__le16 err_csum;
-	__le16 flags;
+	union {
+		struct {
+			u8 reserved[3];
+			u8 flags;
+		}; /* default layout */
+	};
 };
 
-#define ENETC_TXBD_FLAGS_IE	BIT(13)
-#define ENETC_TXBD_FLAGS_F	BIT(15)
+#define ENETC_TXBD_FLAGS_IE	BIT(5)
+#define ENETC_TXBD_FLAGS_F	BIT(7)
 
 union enetc_rx_bd {
 	struct {
-- 
2.17.1

